File: day-6/soln.py
qs = [chr(ord('a')+i) for i in range(26)]

def read_card(ipt):
    card = ''
    for l in ipt:
        if l == '\n':
            yield card
            card = ''
        
        card += l


def part1(ipt):
    total = 0

    for card in read_card(ipt):
        answered = 0

        for q in qs:
            if q in card:
                answered += 1

        total += answered

    return total


def part2(ipt):
    total = 0

    for card in read_card(ipt):
        answered = 0
        counts = {i: 0 for i in qs}

        res = card.split('\n')

        for r in res:
            for q in r:
                counts[q] += 1

        for k, v in counts.items():
            if v == len(res) - 2:
                answered += 1

        total += answered

    return total


if __name__ == '__main__':
    with open('input.txt') as fp:
        print(part1(fp))
        fp.seek(0)
        print(part2(fp))

File: day-6/soln.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func buildAlphabet() []rune {
	var qs []rune

	for i := 'a'; i < 'a' + 26; i++ {
		qs = append(qs, i)
	}

	return qs
}

func readCard(ipt []string, ch chan string) {
	var card strings.Builder
	for _, l := range ipt {
		if l == "" {
			ch <- card.String()
			card.Reset()
			continue
		}

		if strings.HasPrefix(l, " ") {
			card.WriteString(l)
		} else {
			card.WriteString(" ")
			card.WriteString(l)
		}
	}

		ch <- card.String()
		close(ch)
}

func part1(ipt []string) (total int) {
	cards := make(chan string)
	alpha := buildAlphabet()
	go readCard(ipt, cards)

	for card := range cards {
		answered := 0

		for _, q := range alpha {
			if strings.ContainsRune(card, q) {
				answered++
			}
		}

		total += answered
	}

	return
}

func part2(ipt []string) (total int) {
	cards := make(chan string)
	go readCard(ipt, cards)

	for card := range cards {
		answered := 0
		counts := make(map[rune]int)
		res := strings.Split(card, " ")

		for _, r := range res {
			for _, q := range r {
				counts[q]++
			}
		}

		for _, v := range counts {
			if v == len(res) - 1 {
				answered++
			}
		}

		total += answered
	}

	return
}

func main() {
	var ipt []string

	file, _ := os.Open("input.txt")
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		ipt = append(ipt, scanner.Text())
	}

	fmt.Println(part1(ipt))
	fmt.Println(part2(ipt))
}